### DISCUSSION TITLE
Client authentication SPI multiple required authenticators

### DISCUSSION THREAD

**Discussion by bclouser** [keycloak/keycloak#27220] [Q&A]
Hello, 
I have been playing around with creating a custom SPI and adding a custom client authenticator to keycloak. I'm building an authenticator which will communicate with an external service via simple http request to perform non-keycloak verifications. Ultimately i want this authenticator to run **after** the normal `client_credentials` grant flow.

Sadly, this is not supported. I can indeed have multiple authenticators configured, but, multiple authenticators are only considered during an authentication flow if they are all marked as "alternative", which doesn't help me because only one will successfully match the client and authenticate, i need both. If i configure one or more are to be "required" only the first in the list will be executed, which doesn't help me either.

This all happens here: https://github.com/keycloak/keycloak/blob/main/services/src/main/java/org/keycloak/authentication/ClientAuthenticationFlow.java

So my question becomes: **Is having the ability to configure multiple required authenticators (they all must pass to authenticate a request) even make sense to support in keycloak?** I know it would help me out, but i can imagine it isn't common and maybe it's actually quite a bad idea :upside_down_face: .

My current plan is just to nest an instance of `ClientIdAndSecretAuthenticator` inside the class of my custom authenticator and then in my `authenticateClient()` method i first call `clientIdAndSecretAuthenticator.authenticateClient(context);` and check the various results before moving on to performing my custom operations.
I think this should work, but it's still on the operating table so i can't say for sure.

Thanks,
ben

**Comment by danielFesenmeyer** [comment]
I think your use case is possible, but it's a bit cumbersome to configure.

You could try:
Wrap the existing top-level "alternative" elements (Cookie, ..., forms) into a new top-level "required" sub-flow - and provide a meaningful name such as "login alternatives". Then add your authenticator after this new sub-flow, also top-level. Also mark it as "required".
Afaik, the UI currently does not support moving elements, so you have to re-create all those elements manually.
